### PR TITLE
[SPARK-41106][SQL] Reduce collection conversion when create AttributeMap

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr aims to reduce collection conversion when create AttributeMap as following ways:

1. Add a new `apply` method to `AttributeMap`

```
  def apply[A](kvs: Iterable[(Attribute, A)]): AttributeMap[A] = {
    new AttributeMap(kvs.map(kv => (kv._1.exprId, kv)).toMap)
  }
```

and use it in applicable scenarios to avoid additional collection conversion.

Although the new `apply` method is more generic, I did not delete the old ones for forward compatibility.

2. For the following 2 scenarios, `leftStats.attributeStats ++ rightStats.attributeStats` is `AttributeMap ++ AttributeMap`, will return a new `AttributeMap`, so this pr remove the redundant collection conversion.

https://github.com/apache/spark/blob/7d320d784a2d637fd1a8fd0798da3d2a39b4d7cd/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala#L86

https://github.com/apache/spark/blob/7d320d784a2d637fd1a8fd0798da3d2a39b4d7cd/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala#L148

3. For the following scenario, `attributePercentiles` is a `Map` and there is a corresponding `apply` method can accept `Map` input, so remove the redundant `toSeq` in this pr

https://github.com/apache/spark/blob/7d320d784a2d637fd1a8fd0798da3d2a39b4d7cd/sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala#L323

### Why are the changes needed?
Minor performance improvements, reducing collection conversion

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

- Pass GitHub Actions
- Manual test
```
dev/change-scala-version.sh 2.13
build/mvn clean install -Phadoop-3 -Phadoop-cloud -Pmesos -Pyarn -Pkinesis-asl -Phive-thriftserver -Pspark-ganglia-lgpl -Pkubernetes -Phive  -Pscala-2.13 -fn
```

All Test passed

### CODE REVIEW

File: sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala [hunk 1]
```diff
@@ -31,6 +31,10 @@ object AttributeMap {
     new AttributeMap(kvs.map(kv => (kv._1.exprId, kv)).toMap)
   }
 
+  def apply[A](kvs: Iterable[(Attribute, A)]): AttributeMap[A] = {
+    new AttributeMap(kvs.map(kv => (kv._1.exprId, kv)).toMap)
+  }
+
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala:37]:
```
https://github.com/apache/spark/blob/0cb79a4af27eb226a4f7574aa1c8b54412222583/sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala#L26-L32
```
Can we identify the above two as `@ deprecated` now and remove it at an appropriate time(maybe Spark 4.0?)? The new `apply` method is more general.

**Comment by LuciferYang** [comment]
cc @srowen How about this one

File: sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala (hunk 1 above)
Comment (srowen) [reply sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala:37]: I don't think it matters either way; we can keep it or not

**Review by srowen — APPROVED** [review]
Seems fine pending all tests

**Comment by LuciferYang** [comment]
GA passed

**Comment by srowen** [comment]
Cool, so this has passed on 2.12 and 2.13, seems OK to merge

**Comment by LuciferYang** [comment]
Yes, both Scala 2.12 and Scala 2.13 have passed the test :)

**Comment by srowen** [comment]
Merged to master

**Comment by LuciferYang** [comment]
Thanks @srowen ~

**Pull request closed without merging**